File: python/largest-series-product/largest_series_product.py

'''
near plain translation of my php solution, nothing very pythonic
'''
from functools import reduce
import re


def largest_product(series: str, size: int) -> int:
    if series != re.sub(r"/\D/", "", series):
        raise ValueError("Invalid input")

    ser_size = len(series)
    if size > ser_size:
        raise ValueError("Size > series length")

    if not ser_size or not size:
        return 1

    if size < 0:
        raise ValueError("Size must be >= 0")

    hi_product = 0
    for i in range(0, ser_size - size + 1):
        ser_slice = series[i:i+size]
        product = reduce(lambda a, b: int(a)*int(b), ser_slice)
        if product > hi_product:
            hi_product = product

    return hi_product

File: php/largest-series-product/largest-series-product.php
<?php

class Series
{
    public $seq;

    public function __construct(string $seq)
    {
        // seq. must be digits only
        if (preg_match_all("/[^\d]/", $seq, $m)) {
            throw new InvalidArgumentException("Invalid seq.");
        }
        $this->seq = $seq;
    }

    public function largestProduct(int $length): int
    {
        $seq_length = strlen($this->seq);

        // 0 <= length < seq. length
        if ($seq_length < $length || $length < 0) {
            throw new InvalidArgumentException("Invalid length");
        }

        if (!$length) {
            return 1;
        }

        $hi_product = 0;
        $max = $seq_length - $length;

        for ($i = 0; $i <= $max; $i++) {
            $sub = substr($this->seq, $i, $length);
            $product = array_product(str_split($sub));

            if ($product > $hi_product) {
                $hi_product = $product;
            }
        }
        return $hi_product;
    }
}
